src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs

```diff
@@ -184,10 +184,18 @@ private void UpdateSyncModes(SyncMode newModes)
                     if (_logger.IsTrace) _logger.Trace(message);
                 }
             }
-
+            
             SyncMode previous = Current;
+
+            SyncModeChangedEventArgs args = new SyncModeChangedEventArgs(previous, Current);
+            
+            // Changing is invoked here so we can block until all the subsystems are ready to switch
+            // for example when switching to Full sync we need to ensure that we safely transition
+            // the beam sync DB and beam processor
+            Preparing?.Invoke(this, args);
```

Comment:
since these are not async events they allow us to wait for some transitions to happen in three stages and so synchronize the behaviour of beam sync db and beam processor

I spent hours thinking about it and trying various sync mechanisms and this one seems to be relatively easy to follow and neat

Reply:
What is difference between Preparing and Changing?

Reply:
handler handling Changing are sure that all the handlers in the preparing stage have finished executing their transition - just two steps ordering the behaviours